### PR TITLE
feat: Feast Spark Offline Store

### PR DESCRIPTION
**Pull request by kevjumba**
**What this PR does / why we need it**:

Add spark offline store plugin from Adyen Spark to spark offline store. https://github.com/Adyen/feast-spark-offline-store
Currently 159/194 integration tests for spark source work. 

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
Initial cut of Spark based Feast offline store
```

### CODE REVIEW

File: sdk/python/feast/type_map.py [hunk 1]
```diff
@@ -326,13 +326,17 @@ def _python_value_to_proto_value(
...
+            val_list = []
+            for value in values:
+                if value.dtype == "bool":
+                    value = [bool(e) for e in value]
+                    val_list.append(ProtoValue(**{field_name: proto_type(val=value)}))  # type: ignore
+                elif value is not None:
+                    val_list.append(ProtoValue(**{field_name: proto_type(val=value)}))  # type: ignore
+                else:
+                    val_list.append(ProtoValue())  # type: ignore
+            return val_list
```
Comment (kevjumba) [inline sdk/python/feast/type_map.py:339]: FYI, I added this in to get some of my tests to pass but I'm not 100% confident that it is actually correct for the larger codebase. Running against the integration tests on master using Github CI to see if it breaks anything. JIC this looks suspicious. same below

File: sdk/python/feast/type_map.py [hunk 2]
```diff
@@ -353,9 +357,14 @@ def _python_value_to_proto_value(
...
-
             return [
-                ProtoValue(**{field_name: func(value)})
+                ProtoValue(
+                    **{
+                        field_name: func(
+                            bool(value) if type(value) is np.bool_ else value  # type: ignore
+                        )
+                    }
+                )
```
Comment (kevjumba) [inline sdk/python/feast/type_map.py:367]: same here.

File: sdk/python/feast/infra/offline_stores/third_party/spark.py [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+        query = offline_utils.build_point_in_time_query(
+            feature_view_query_contexts=query_context,
+            left_table_query_string=tmp_entity_df_table_name,
+            entity_df_event_timestamp_col=event_timestamp_col,
+            entity_df_columns=entity_schema.keys(),
+            query_template=MULTIPLE_FEATURE_VIEW_POINT_IN_TIME_JOIN,
+            full_feature_names=full_feature_names,
+        )
+
+        # TODO: Figure out what this is used for
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark.py:152]: Remove this?

File: sdk/python/feast/infra/offline_stores/third_party/spark.py [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+        event_timestamp_column: str,
+        start_date: datetime,
+        end_date: datetime,
+    ) -> RetrievalJob:
+        """
+        Note that join_key_columns, feature_name_columns, event_timestamp_column, and created_timestamp_column
+        have all already been mapped to column names of the source table and those column names are the values passed
+        into this function.
+        """
+        return SparkOfflineStore.pull_latest_from_table_or_query(
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark.py:187]: This seems weird? Other offline stores create a different query that generates all feature values between a time range. Any reason why this is implemented differently?

File: sdk/python/feast/infra/offline_stores/third_party/spark_source.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+    def __init__(
+        self,
+        table: Optional[str] = None,
+        query: Optional[str] = None,
+        # TODO support file readers
+        # path: Optional[str] = None,
+        # jdbc=None,
+        # format: Optional[str] = None,
+        # options: Optional[Dict[str, Any]] = None,
+        event_timestamp_column: Optional[str] = None,
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_source.py:28]: feast today does some event timestamp inference in https://github.com/feast-dev/feast/blob/master/sdk/python/feast/inference.py#L99-L99

Do we need to support this somehow too? it basically checks if there is only one timestamp column

File: sdk/python/feast/type_map.py (hunk 1 above)
Comment (adchia) [reply sdk/python/feast/type_map.py:339]: Yeah def looks suspicious.  Seems like it should maybe be fixed in PYTHON_LIST_VALUE_TYPE_TO_PROTO_VALUE

File: sdk/python/tests/integration/registration/test_universal_types.py [hunk 6]
```diff
@@ -184,9 +184,9 @@ def test_feature_get_historical_features_types_match(offline_types_test_fixtures
     historical_features = fs.get_historical_features(
         entity_df=entity_df, features=features,
     )
+
     # Note: Pandas doesn't play well with nan values in ints. BQ will also coerce to floats if there are NaNs
     historical_features_df = historical_features.to_df()
-    print(historical_features_df)
```
Comment (adchia) [inline sdk/python/tests/integration/registration/test_universal_types.py:189]: revert these changes? below too

File: sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    def create_data_source(
+        self,
+        df: pd.DataFrame,
+        destination_name: str,
+        event_timestamp_column="ts",
+        created_timestamp_column="created_ts",
+        field_mapping: Dict[str, str] = None,
+        **kwargs,
+    ) -> DataSource:
+        # df["event_timestamp"] = pd.to_datetime(df["event_timestamp"], utc=True)
```
Comment (adchia) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py:68]: Is this needed?

File: sdk/python/feast/infra/offline_stores/third_party/spark.py [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+    def get_historical_features(
+        config: RepoConfig,
+        feature_views: List[FeatureView],
+        feature_refs: List[str],
+        entity_df: Union[pandas.DataFrame, str],
+        registry: Registry,
+        project: str,
+        full_feature_names: bool = False,
+    ) -> RetrievalJob:
+        assert isinstance(config.offline_store, SparkOfflineStoreConfig)
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark.py:108]: throw warnings in each of the methods calling out this is alpha?

File: java/datatypes/src/main/proto/feast/core/DatastoreTable.proto [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+package feast.core;
+option java_package = "feast.proto.core";
+option java_outer_classname = "DatastoreTableProto";
+option go_package = "github.com/feast-dev/feast/sdk/go/protos/feast/core";
+
+import "google/protobuf/wrappers.proto";
+
+// Represents a Datastore table
+message DatastoreTable {
```
Comment (pyalex) [inline java/datatypes/src/main/proto/feast/core/DatastoreTable.proto:27]: @kevjumba it seems that this shouldn't be here. can you rebase with latest master please?

File: docs/reference/data-sources/spark.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
+# Snowflake
```
Comment (adchia) [inline docs/reference/data-sources/spark.md:1]: Spark?

File: docs/reference/offline-stores/README.md [hunk 11]
```diff
@@ -9,3 +9,5 @@ Please see [Offline Store](../../getting-started/architecture-and-components/off
 {% page-ref page="bigquery.md" %}
 
 {% page-ref page="redshift.md" %}
+
+{% page-ref page="snowflake.md" %}
```
Comment (adchia) [inline docs/reference/offline-stores/README.md:13]: spark?

File: docs/reference/offline-stores/spark.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+## Description
+
+The Spark offline store is an offline store currently in alpha development that provides support for reading [SparkSources](../data-sources/spark.md).
+
+## Disclaimer
+
+This Spark offline store still does not achieve full test coverage and continues to fail some integration tests when integrating with the feast universal test suite. Please do NOT assume complete stability of the API.
+
+As of 3/1/2022, 179/194 integration tests pass.
```
Comment (adchia) [inline docs/reference/offline-stores/spark.md:11]: Probably can just omit this

File: docs/reference/offline-stores/spark.md [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+* Spark tables and views are allowed as sources that are loaded in from some Spark store(e.g in Hive or in memory).
+* Entity dataframes can be provided as a SQL query or can be provided as a Pandas dataframe. Pandas dataframes will be converted to a Spark dataframe and processed as a temporary view.
+* A `SparkRetrievalJob` is returned when calling `get_historical_features()`.
+  * This allows you to call
+     * `to_df` to retrieve the pandas dataframe.
+     * `to_arrow` to retrieve the dataframe as a pyarrow Table.
+     * `to_spark-df` to retrieve the dataframe the spark.
```
Comment (adchia) [inline docs/reference/offline-stores/spark.md:19]: to_spark_df seems more conventional?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 14]
```diff
@@ -0,0 +1,555 @@
...
+    ) -> RetrievalJob:
+        spark_session = get_spark_session_or_start_new_with_repoconfig(
+            config.offline_store
+        )
+        assert isinstance(config.offline_store, SparkOfflineStoreConfig)
+        assert isinstance(data_source, SparkSource)
+
+        warnings.warn(
+            "The spark offline store is an experimental feature in alpha development. "
+            "This API is unstable and it could and most probably will be changed in the future.",
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:65]: The API is probably stable, so maybe remove that?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 15]
```diff
@@ -0,0 +1,555 @@
...
+    ):
+        super().__init__()
+        self.spark_session = spark_session
+        self.query = query
+        self._full_feature_names = full_feature_names
+        self._on_demand_feature_views = on_demand_feature_views
+        self._metadata = metadata
+
+    @property
+    def full_feature_names(self) -> bool:
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:244]: we've started moving towards just public attributes, so maybe remove these and just do self.metadata = metadata in the constructor, etc?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 16]
```diff
@@ -0,0 +1,555 @@
...
+    store_config: SparkOfflineStoreConfig,
+) -> SparkSession:
+    spark_session = SparkSession.getActiveSession()
+    if not spark_session:
+        spark_builder = SparkSession.builder
+        spark_conf = store_config.spark_conf
+        if spark_conf:
+            spark_builder = spark_builder.config(
+                conf=SparkConf().setAll([(k, v) for k, v in spark_conf.items()])
+            )  # noqa
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:292]: why is this noqa needed?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 17]
```diff
@@ -0,0 +1,555 @@
...
+
+        spark_session = spark_builder.getOrCreate()
+    spark_session.conf.set(
+        "spark.sql.parser.quotedRegexColumnNames", "true"
+    )  # important!
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:297]: remove the comment?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 18]
```diff
@@ -0,0 +1,555 @@
...
+            )
+        entity_df_event_timestamp_range = (
+            entity_df_event_timestamp.min().to_pydatetime(),
+            entity_df_event_timestamp.max().to_pydatetime(),
+        )
+    elif isinstance(entity_df, str):
+        # If the entity_df is a string (SQL query), determine range
+        # from table
+        df = spark_session.sql(entity_df).select(entity_df_event_timestamp_col)
+        entity_df_event_timestamp_range = (
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:322]: you probably need a utc conversion here, but can leave as a TODO?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py [hunk 19]
```diff
@@ -0,0 +1,555 @@
...
+    entity_df: Union[pd.DataFrame, str],
+    entity_df_event_timestamp_col: str,
+    feature_refs: List[str],
+    feature_views: List[FeatureView],
+    spark_session: SparkSession,
+    table_name: str,
+    registry: Registry,
+    project: str,
+) -> List[FeatureViewQueryContext]:
+    # interface of offline_utils.get_feature_view_query_context changed in feast==0.17
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:371]: remove this?

Comment (adchia) [reply sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark.py:371]: i don't see where you use this method?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+                raise ValueError(
+                    f"'file_format' should be one of {self.allowed_formats}"
+                )
+
+        self._spark_options = SparkOptions(
+            table=table, query=query, path=path, file_format=file_format,
+        )
+
+    @property
+    def spark_options(self):
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:72]: similarly, instead of using property, prefer public fields instead

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 21]
```diff
@@ -0,0 +1,307 @@
...
+    @property
+    def file_format(self):
+        """
+        Returns the file format of this feature data source.
+        """
+        return self._spark_options.file_format
+
+    @staticmethod
+    def from_proto(data_source: DataSourceProto) -> Any:
+
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:115]: remove extra line?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+        df = spark_session.sql(f"SELECT * FROM {self.get_table_query_string()}")
+        return (
+            (fields["name"], fields["type"])
+            for fields in df.schema.jsonValue()["fields"]
+        )
+
+    def get_table_query_string(self) -> str:
+        """Returns a string that can directly be used to reference this table in SQL"""
+        if self.table:
+            return f"`{self.table}`"
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:169]: can you leave a comment on why you need to have the quotes (`) ?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+        if self.query:
+            return f"({self.query})"
+
+        # If both the table query string and the actual query are null, we can load from file.
+        spark_session = SparkSession.getActiveSession()
+        if spark_session is None:
+            raise AssertionError("Could not find an active spark session.")
+        df = spark_session.read.format(self.file_format).load(self.path)
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:177]: should you catch any exception here (e.g. it isn't a file) so we can print a more helpful message in the exception?

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 24]
```diff
@@ -0,0 +1,307 @@
...
+        query: Optional[str] = None,
+        path: Optional[str] = None,
+        file_format: Optional[str] = None,
+    ):
+        self._table = table
+        self._query = query
+        self._path = path
+        self._file_format = file_format
+
+    @property
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:198]: property vs public field again

File: sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py [hunk 25]
```diff
@@ -0,0 +1,307 @@
...
+    _proto_attr_name = "spark_storage"
+
+    spark_options: SparkOptions
+
+    def __init__(self, table_ref: Optional[str] = None, query: Optional[str] = None):
+        self.spark_options = SparkOptions(table=table_ref, query=query)
+
+    @staticmethod
+    def from_proto(storage_proto: SavedDatasetStorageProto) -> SavedDatasetStorage:
+        # options = SparkOptions.from_proto(
```
Comment (adchia) [inline sdk/python/feast/infra/offline_stores/third_party/spark_offline_store/spark_source.py:297]: Leave TODOs here where you'll have future work?

File: sdk/python/feast/type_map.py (hunk 1 above)
Comment (adchia) [reply sdk/python/feast/type_map.py:341]: would rather keep it in its original state and leaving maybe a TODO in that PYTHON_LIST_VALUE_TYPE_TO_PROTO_VALUE block to fix boolean lists.

File: sdk/python/feast/type_map.py [hunk 26]
```diff
@@ -566,3 +577,51 @@ def _non_empty_value(value: Any) -> bool:
     return value is not None and (
         not isinstance(value, Sized) or len(value) > 0 or isinstance(value, str)
     )
+
+
+def spark_to_feast_value_type(spark_type_as_str: str) -> ValueType:
+    # TODO not all spark types are convertible
```
Comment (adchia) [inline sdk/python/feast/type_map.py:583]: can you give some examples of types that can't be converted?

File: sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+    def teardown(self):
+        self.spark_session.stop()
+
+    def create_offline_store_config(self):
+        self.spark_offline_store_config = SparkOfflineStoreConfig()
+        self.spark_offline_store_config.type = "spark"
+        self.spark_offline_store_config.spark_conf = self.spark_conf
+        return self.spark_offline_store_config
+
+    # abstract
```
Comment (adchia) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py:58]: remove comment

File: sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+            timestamp_mapping = {value: key for key, value in field_mapping.items()}
+            if (
+                event_timestamp_column in timestamp_mapping
+                and timestamp_mapping[event_timestamp_column] in df
+            ):
+                col = timestamp_mapping[event_timestamp_column]
+                df[col] = pd.to_datetime(df[col], utc=True)
+
+        # https://stackoverflow.com/questions/51871200/analysisexception-it-is-not-allowed-to-add-database-prefix
+        # destination_name = self.get_prefixed_table_name(destination_name)
```
Comment (adchia) [inline sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py:83]: can you instead just prepend the project + project name?

File: sdk/python/tests/integration/feature_repos/repo_configuration.py [hunk 29]
```diff
@@ -52,7 +55,7 @@
     "type": "redis",
     "redis_type": "redis_cluster",
     # Redis Cluster Port Forwarding is setup in "pr_integration_tests.yaml" under "Setup Redis Cluster".
-    "connection_string": "127.0.0.1:6001,127.0.0.1:6002,127.0.0.1:6003",
+    "connection_string": "127.0.0.1:32001,127.0.0.1:32002,127.0.0.1:32003",
```
Comment (adchia) [inline sdk/python/tests/integration/feature_repos/repo_configuration.py:58]: seems like a different PR?

File: sdk/python/tests/integration/feature_repos/repo_configuration.py [hunk 30]
```diff
@@ -74,6 +77,9 @@
             IntegrationTestRepoConfig(online_store=REDIS_CONFIG),
             IntegrationTestRepoConfig(online_store=REDIS_CLUSTER_CONFIG),
             # GCP configurations
+            IntegrationTestRepoConfig(
```
Comment (adchia) [inline sdk/python/tests/integration/feature_repos/repo_configuration.py:80]: probably would not include them here. we can make a separate make command that tests third party stores too though.

File: sdk/python/feast/type_map.py (hunk 26 above)
Comment (kevjumba) [reply sdk/python/feast/type_map.py:594]: Done.

**Review by kevjumba — COMMENTED** [review]
@adchia addressed changes PTAL!

File: sdk/python/feast/type_map.py [hunk 31]
```diff
@@ -326,7 +328,14 @@ def _python_value_to_proto_value(
                     ProtoValue(unix_timestamp_list_val=Int64List(val=ts))  # type: ignore
                     for ts in int_timestamps_lists
                 ]
-
+            if feast_value_type == ValueType.BOOL_LIST:
+                # ProtoValue no longer supports conversion of np.bool_ so we need it convert it.
```
Comment (adchia) [inline sdk/python/feast/type_map.py:332]: did it support it at some point?

File: sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py (hunk 28 above)
Comment (adchia) [reply sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py]: right but can we e.g. not have periods to denote they're databases? it seems reasonable to at least ensure that the name is unique

File: sdk/python/feast/type_map.py (hunk 31 above)
Comment (kevjumba) [reply sdk/python/feast/type_map.py:332]: I'm not quite sure, I couldn't find any real information on it but it was in the codebase and was working, ProtoValue just doesn't work on np.bool_ now.

Comment (adchia) [reply sdk/python/feast/type_map.py:332]: Got it. I'd update the comment that it's to support np.bool_ instead then

Comment (kevjumba) [reply sdk/python/feast/type_map.py:332]: Done.

File: sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py (hunk 28 above)
Comment (kevjumba) [reply sdk/python/tests/integration/feature_repos/universal/data_sources/spark_data_source_creator.py]: Fixed, we can't have periods but we can use underscores to prepend.

**Review by adchia — APPROVED** [review]
/lgtm

File: sdk/python/feast/infra/offline_stores/contrib/spark_offline_store/spark.py [hunk 32]
```diff
@@ -0,0 +1,522 @@
...
+        end_date_str = _format_datetime(end_date)
+        query = f"""
+                SELECT
+                    {field_string}
+                    {f", {repr(DUMMY_ENTITY_VAL)} AS {DUMMY_ENTITY_ID}" if not join_key_columns else ""}
+                FROM (
+                    SELECT {field_string},
+                    ROW_NUMBER() OVER({partition_by_join_key_string} ORDER BY {timestamp_desc_string}) AS feast_row_
+                    FROM {from_expression} t1
+                    WHERE {event_timestamp_column} BETWEEN TIMESTAMP('{start_date_str}') AND TIMESTAMP('{end_date_str}')
```
Comment (rehevkor5) [inline sdk/python/feast/infra/offline_stores/contrib/spark_offline_store/spark.py:92]: I'm not sure if it was intentional that both the start and the end are inclusive, but IMO the starting point should be inclusive and the ending point should be exclusive. That way, if you have ranges of tuples like (a, b) and (b, c), they are contiguous and non-overlapping. By using `BETWEEN` the ranges will instead overlap at `b`.
